File: examples/c/shapes.c
/* Real C structs: value declaration with a brace initializer, pointer params
 * (`struct Point *p` is the IR's reference), -> and . field access, and
 * mutation through a function. All 7 other targets match the original C. */
#include <stdio.h>

struct Point {
    int x;
    int y;
};

void translate(struct Point *p, int dx, int dy) {
    p->x += dx;
    p->y += dy;
}

int manhattan(struct Point *p) {
    int ax = p->x;
    int ay = p->y;
    if (ax < 0) {
        ax = -ax;
    }
    if (ay < 0) {
        ay = -ay;
    }
    return ax + ay;
}

int main(void) {
    struct Point p = {3, 4};
    printf("x: %d\n", p.x);
    printf("y: %d\n", p.y);
    translate(&p, 10, -9);
    printf("x: %d\n", p.x);
    printf("y: %d\n", p.y);
    printf("dist: %d\n", manhattan(&p));
    p.y++;
    printf("y: %d\n", p.y);
    return 0;
}

File: examples/py/shapes.py
# Real Python classes become IR structs: field types are inferred from the
# __init__ call sites, and mutation through a function works on every target.
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

class Tally:
    def __init__(self, label, count):
        self.label = label
        self.count = count

def translate(p, dx, dy):
    p.x += dx
    p.y += dy

def bump(t, n):
    t.count += n
    t.label = t.label + "!"

def main():
    p = Point(2, 3)
    translate(p, 5, -1)
    print(p.x)
    print(p.y)
    print(p.x * p.y)
    t = Tally("run", 1)
    bump(t, 4)
    bump(t, 5)
    print(t.label)
    print(t.count)

main()
